### ISSUE TITLE
Rule Change: Add self to global objects checked by no-eval

### ISSUE THREAD

**Issue by sethamus** [eslint/eslint#20053]
### What rule do you want to change?

no-eval

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/*eslint no-eval: "error"*/
/*global self*/

self.eval("const a = 0");
```

### What does the rule currently do for this code?

Currently, the rule does not flag the use of eval when accessed via the self global object. Only window, global, and globalThis are checked.

### What will the rule do after it's changed?

After the change, the rule will also flag eval when accessed via the self global object, in addition to the existing global objects. For example, the code above will be reported as a violation.

### Participation

- [x] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Makes sense to me 👍 

Some other core rules are already treating `self` as a global object. For example, `prefer-regex-literals` ([Playground Demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgcHJlZmVyLXJlZ2V4LWxpdGVyYWxzOiBcImVycm9yXCIqL1xuLypnbG9iYWwgc2VsZiovXG5cbm5ldyBzZWxmLlJlZ0V4cChcImFcIik7XG4gICIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==)).

@eslint/eslint-team thoughts?

**Comment by fasttime** [comment]
I think it would be good to have a central way to specify names of the global object across rules. There are several rules all with their own idea of how the global object can be named, and it's confusing.

| | `global` | `globalThis` | `self` | `window` | custom names |
| :- | :-: | :-: | :-: | :-: | :-: |
| `no-alert` | | ✗ | | ✗ | |
| `no-eval` | ✗ | ✗ | | ✗ | |
| `no-implied-eval` | ✗ | ✗ | | ✗ | |
| `no-restricted-globals` | | ✗ | ✗ | ✗ | ✗ |

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZ2xvYmFsIGdsb2JhbCwgc2VsZiwgd2luZG93LCBjdXN0b21HbG9iYWwgKi9cblxuLyplc2xpbnQgbm8tYWxlcnQ6IFwiZXJyb3JcIiovXG5cbmdsb2JhbC5hbGVydChcIlwiKTtcbmdsb2JhbFRoaXMuYWxlcnQoXCJcIik7XG5zZWxmLmFsZXJ0KFwiXCIpO1xud2luZG93LmFsZXJ0KFwiXCIpO1xuXG4vKmVzbGludCBuby1ldmFsOiBcImVycm9yXCIqL1xuXG5nbG9iYWwuZXZhbChcIlwiKTtcbmdsb2JhbFRoaXMuZXZhbChcIlwiKTtcbnNlbGYuZXZhbChcIlwiKTtcbndpbmRvdy5ldmFsKFwiXCIpO1xuXG4vKmVzbGludCBuby1pbXBsaWVkLWV2YWw6IFwiZXJyb3JcIiovXG5cbmdsb2JhbC5zZXRUaW1lb3V0KFwiXCIpO1xuZ2xvYmFsVGhpcy5zZXRUaW1lb3V0KFwiXCIpO1xuc2VsZi5zZXRUaW1lb3V0KFwiXCIpO1xud2luZG93LnNldFRpbWVvdXQoXCJcIik7XG5cbi8qIGVzbGludCBuby1yZXN0cmljdGVkLWdsb2JhbHM6IFtcImVycm9yXCIsIHsgZ2xvYmFsczogW1wicmVzdHJpY3RlZFwiXSwgY2hlY2tHbG9iYWxPYmplY3Q6IHRydWUsIGdsb2JhbE9iamVjdHM6IFtcImN1c3RvbUdsb2JhbFwiXSB9XSAqL1xuXG5nbG9iYWwucmVzdHJpY3RlZDtcbmdsb2JhbFRoaXMucmVzdHJpY3RlZDtcbnNlbGYucmVzdHJpY3RlZDtcbndpbmRvdy5yZXN0cmljdGVkO1xuY3VzdG9tR2xvYmFsLnJlc3RyaWN0ZWQ7XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=)

Many other rules like the already mentioned `prefer-regex-literals`, use [`ReferenceTracker`](https://eslint-community.github.io/eslint-utils/api/scope-utils.html#referencetracker-class) with the default list of global names, which defaults to `global`, `globalThis`, `self` and `window`.

**Comment by sethamus** [comment]
@fasttime, I was actually working on #19950, and what I did was I added a shared constant for global object names at `lib/rules/utils/global-object-names.js`:

```js
"use strict";
module.exports = Object.freeze(["global", "globalThis", "self", "window"]);
```

I then refactored the existing rules to import this constant instead of hardcoding the global object names. What do you think of this approach?

**Comment by nzakas** [comment]
Can we merge this issue into https://github.com/eslint/eslint/issues/19950 so we don't have multiple issues related to this?
